### ISSUE TITLE
Extended tap formatter gets confused when retrying  tests

### ISSUE THREAD

**Issue by jacksgt** [bats-core/bats-core#709]
Hello everyone!

Thanks for this excellent and useful project!

**Describe the bug**
I believe I found a bug when in the extended tap parser that gets triggered when tests are re-tried (with the recently introduced `BATS_TEST_RETRIES` feature).

**To Reproduce**

Test file:

```sh
#!/usr/bin/env bats

BATS_TEST_RETRIES=1

@test "test foobar" {
  echo "Hello, world!"
  exit 1
}
```

Run with `bats -p test-file.bats`.

Observe strange `begin 1` or `begin 2` output in front of the test name.

```sh
# bats -F pretty test-file.bats 
test-file.bats
[1/1]   test foobar
2 begin 1 test foobar
[2/1]   begin 1 test foobar
 ✗
   (in test file test-file.bats, line 7)
     `exit 1' failed
   Hello, world!

1 test, 1 failure
```

**Expected behavior**
All extended formatters should handle retried tests correctly.

**Environment (please complete the following information):**
 - Bats Version: 1.9.0
 - OS: Alma 9 Linux
 - Bash version: 5.1.8

**Additional context**

I believe the issue is coming from this line: https://github.com/bats-core/bats-core/blob/e4ddb10402d9df809396938f48b36ae73d10dcd6/lib/bats-core/formatter.bash#L54

Basically, the parser assumes that `begin_index` will always match `index` (and thus can be used to strip off the number from the line). However, this is not the case for retried tests, since bats does not increment the `index` for each test retry, but `begin_index` is incremented for each `begin` tap line.

This can be seen based on the following `--formatter cat` output (where the test `If namespace has no label then route will get external-dns annotation pointing to default router shard` was retried twice (and failed twice).

```sh
# bats -F cat test-file.bats 
1..1
suite /okd4-install/test-file.bats
begin 1 test foobar
begin 1 test foobar
not ok 1 test foobar
# (in test file test-file.bats, line 7)
#   `exit 1' failed
# Hello, world!
```

**Comment by jacksgt** [comment]
A fix (albeit non-robust) could look like this:

```diff
    'begin '*) # this might only be called in extended tap output
      ((++begin_index))
      scope=begin
      # example line: "begin 42 my-foobar-test"
-     test_name="${line#* "$begin_index" }"
+     test_name=$(echo "${line} | cut -d' ' -f3-")
      bats_tap_stream_begin "$begin_index" "$test_name"
      ;;
```

This will correctly strip off the leading "begin {index+retry}".

**Comment by martin-schulze-vireso** [comment]
I could not reproduce this. However, the observed behavior looks like something I have seen before: Sometimes the `begin`s leak into the pretty output. This does not seem to be related to retried tests though.

Was this problem reliably reproducible with your code above?

**Comment by jacksgt** [comment]
You are right, with this minimal working example (from scratch) I can not reproduce the issue:

```sh
podman run --rm -it docker.io/almalinux/9-base bash

yum install -y gettext ncurses

BATS_VERSION=1.9.0
BATS_SHA256=3c64295318d157616ca2478de819371e2c579ccdab30c1f0b7279f7b214fb1f4
cd /tmp
curl -sL "https://github.com/bats-core/bats-core/archive/refs/tags/v${BATS_VERSION}.tar.gz" -o /tmp/bats-core.tar.gz
echo "${BATS_SHA256}  /tmp/bats-core.tar.gz" | sha256sum -c -
tar xzvf bats-core.tar.gz && \
./bats-core-${BATS_VERSION}/install.sh /usr/local

cat - > test-file-1.bats <<EOF
#!/usr/bin/env bats

BATS_TEST_RETRIES=1

@test "test not ok" {
  echo "Hello, world!"
  exit 1
}
EOF

cat - > test-file-2.bats <<EOF
#!/usr/bin/env bats

BATS_TEST_RETRIES=1

@test "test ok" {
  echo "Hello, world!"
}
EOF

bats -p *.bats
test-file-1.bats
 ✗ test not ok
   (in test file test-file-1.bats, line 7)
     `exit 1' failed
   Hello, world!
test-file-2.bats
 ✓ test ok

2 tests, 1 failure
```

I will keep digging to see if I can reproduce it in another way...

**Comment by jacksgt** [comment]
It definitely looks like there is something wrong with the parsing logic of `bats_parse_internal_extended_tap`. 
I'm not sure if you can call this a "reproducer", but I have put together the following "dumb" formatter:

```sh
#!/usr/bin/env bash

set -e

# shellcheck source=lib/bats-core/formatter.bash
source "$BATS_ROOT/lib/bats-core/formatter.bash"

bats_tap_stream_plan() {
  echo "STREAM_PLAN: $1"
}

bats_tap_stream_begin() {
  echo "STREAM_BEGIN: index: ${1}, name: ${2}"
}

bats_tap_stream_ok() {
  echo "STREAM_OK: index ${1}, name: ${2}"
}

bats_tap_stream_skipped() {
  echo "STREAM_SKIPPED: index ${1}, name ${2}, reason ${3}"
}

bats_tap_stream_not_ok() {
  echo "STREAM_NOT_OK: index ${1}, name ${2}"
}

bats_tap_stream_comment() { # <comment> <scope>
  echo "STREAM_COMMENT: comment ${1}, scope ${2}"
}

bats_tap_stream_suite() {
  echo "STREAM_SUITE: file ${1}"
}

bats_tap_stream_unknown() { # <full line> <scope>
  echo "STREAM_UNKNOWN: ${1}, scope: ${2}"
}

bats_parse_internal_extended_tap
```

(this is basically a stripped down version of the "pretty" formatter).

If I run the tests mentioned in my previous comment, I get the following output:
```sh
# bats -p *.bats
STREAM_PLAN: 2
STREAM_SUITE: file /tmp/test-file-1.bats
STREAM_BEGIN: index: 1, name: test not ok
STREAM_BEGIN: index: 2, name: begin 1 test not ok
STREAM_NOT_OK: index 1, name test not ok
STREAM_COMMENT: comment (in test file test-file-1.bats, line 7), scope not_ok
STREAM_COMMENT: comment   `exit 1' failed, scope not_ok
STREAM_COMMENT: comment Hello, world!, scope not_ok
STREAM_SUITE: file /tmp/test-file-2.bats
STREAM_BEGIN: index: 3, name: begin 2 test ok
STREAM_OK: index 2, name: test ok
```

Note especially these lines:
```
STREAM_BEGIN: index: 1, name: test not ok
STREAM_BEGIN: index: 2, name: begin 1 test not ok
```
the first `bats_tap_stream_begin` invocation correctly receives `test not ok` as the name of the test, but the second invocation receives `begin test not ok`.
I believe this is due to the logic I mentioned in my original post: https://github.com/bats-core/bats-core/blob/e4ddb10402d9df809396938f48b36ae73d10dcd6/lib/bats-core/formatter.bash#L54
